### DISCUSSION TITLE
Conversion to TFlite

### DISCUSSION THREAD

**Discussion by nahla-naz** [shubham0204/OnDevice-Face-Recognition-Android#6] [Q&A]
Hi,

**How do you convert models to TFlite?** I tried converting, but my app is crashing while trying to get output.

Do you perform optimisation/quantisation? If you do, does it reduce the model accuracy?

Please help. Thanks.

**Comment by shubham0204** [comment]
You can download the H5 file from the URL present in this script: https://github.com/serengil/deepface/blob/master/deepface/basemodels/Facenet.py

and use the `tf.lite.TFLiteConverter.from_keras_model` method to convert the model to TFLite. It also provides support for `int8` and `float16` quantization. I haven't checked the impact of quantization on accuracy (the app uses non-quantized models).

**Comment by nahla-naz** [comment]
Hi @shubham0204 , Thank you for your prompt reply. 

I'm trying to convert` FaceNet` model from `keras_facenet` python library.

I am able to convert it using Tensorflow 2.17 and get output in python. But it doesn't run on android because latest version for TFlite in android is 2.16.1. I tried converting the model using lower versions of Tensorflow in python, but the models don't work.

I am getting this error while trying to run TFlite model on android: 
_Caused by: java.lang.IllegalStateException: Internal error: Unexpected failure when preparing tensor allocations: tensorflow/lite/core/subgraph.cc **BytesRequired number of elements overflowed.**_


- I used the same method as you mentioned to convert the model to TFlite. 
- I checked TFlite documentation regarding accuracy loss : there is small accuracy loss based on which quantisation method you have used - sometimes the accuracy loss is insignificant.

**Reply by shubham0204** [reply]
Considering the error message, it does not seem to be a problem arising from incorrect model conversion but rather from an incorrect shape given for any input/output type on the Android side. You can check the [input and output shapes of the TFLite model](https://www.tensorflow.org/lite/guide/inference#load_and_run_a_model_in_python) and compare them with the ones given in your Kotlin/Android code.

**Reply by nahla-naz** [reply]
@shubham0204 Will check this. However, I think the error is occurring while loading the model before passing any input.

**Comment by dengwhao** [comment]
I'm curious—why is your converted model only 23 MB, while my FP32-converted model is 89 MB? I also downloaded the original FaceNet512 H5 model, and it appears to be around 89 MB as well. Could you please explain this discrepancy?
